### PR TITLE
sgx-sdk: 2.16 -> 2.21

### PR DESCRIPTION
**Pull request by phlip9**
## Description of changes

This PR updates the Intel SGX SDK and AESM service packages to the latest upstream release: <https://github.com/intel/linux-sgx/releases/tag/sgx_2.21>.

I've tested everything, up to and including successful remote attestion on an Azure gen2 VM running NixOS with the NixOS `aesmd` service.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

After testing again, the NixOS `aesmd` service appears to work just fine with the `azure-dcap-client`. Ignore the section below (left for posterity):

<details>

* NixOS `aesmd`
  * By itself, the existing NixOS `aesmd` service seems to still work for launching enclaves and doing local attestation.
  * However, I couldn't get the existing NixOS `aesmd` service to work with the azure DCAP provider. OTOH a simple service config w/o the dynamic user or sandboxing does work. I'm not super familiar with systemd's sandboxing or how to debug it, so I'll probably stick with what I got for now. I'll include the service config at the end for reference.

```nix
{ pkgs, ... }:
{
  hardware.cpu.intel.sgx.provision.enable = true;
  hardware.cpu.intel.sgx.enableDcapCompat = true;

  users.groups.aesmd = {};
  users.users.aesmd = {
    description = "User for Intel Architectural Enclave Service Manager daemon";
    isSystemUser = true;
    createHome = true;
    group = "aesmd";
    extraGroups = ["sgx" "sgx_prv"];
    home = "/var/opt/aesmd";
  };

  systemd.services.aesmd = let
    storeAesmFolder = "${pkgs.sgx-psw}/aesm";
    quoteProviderLibrary = pkgs.sgx-azure-dcap-client;
  in {
    description = "Intel Architectural Enclave Service Manager";
    wantedBy = ["multi-user.target"];

    after = [
      "auditd.service"
      "network.target"
      "syslog.target"
    ];

    environment = {
      NAME = "aesm_service";
      AESM_PATH = storeAesmFolder;
      LD_LIBRARY_PATH = lib.makeLibraryPath [
        quoteProviderLibrary
        storeAesmFolder
      ];
      AZDCAP_DEBUG_LOG_LEVEL = "INFO";
      AZDCAP_CACHE = config.users.users.aesmd.home;
    };

    serviceConfig = {
      Type = "simple";
      User = "aesmd";
      Group = "aesmd";

      WorkingDirectory = storeAesmFolder;

      ExecStartPre = let
        preStartScript = pkgs.writeShellScript "aesmd-pre-start.sh" ''
          set -eux

          ${pkgs.coreutils}/bin/mkdir -p /var/run/aesmd/
          ${pkgs.coreutils}/bin/chown -R aesmd:aesmd /var/run/aesmd/
          ${pkgs.coreutils}/bin/chmod 0755 /var/run/aesmd/

          ${pkgs.coreutils}/bin/chmod 0755 /var/opt/
        '';
      in
        "+${preStartScript}";

      ExecStart = "${pkgs.sgx-psw}/bin/aesm_service --no-daemon --supported_attestation_types=ECDSA";
      ExecReload = ''+${pkgs.coreutils}/bin/kill -SIGHUP "$MAINPID"'';

      Restart = "on-failure";
      RestartSec = "15s";
    };
  };

  nixpkgs.config.permittedInsecurePackages = [
    "openssl-1.1.1v"
  ];
}
```

</details>

### CODE REVIEW

**Comment by phlip9** [comment]
I'm not sure why `ofborg-eval` is erroring out... going to try bisecting here

**Comment by phlip9** [comment]
Looks like `ofborg` is skipping the automated tests b/c the nixpkgs config needs `permittedInsecurePackages = [ "openssl-1.1.1w" ];` or `NIXPKGS_ALLOW_INSECURE=1`. Is there some way to add one of these to the `ofborg` eval?

**Comment by sbellem** [comment]
Is there a way we could first merge https://github.com/NixOS/nixpkgs/pull/236815, so that:

* we get smaller diffs
* people who may need intermediate versions can have them
* done work, is not lost

**Comment by phlip9** [comment]
> Is there a way we could first merge #236815

Sure, I'm fine either way. Doubtful that the diff will be appreciably smaller, though. At least rebasing won't be very difficult 😅

It seems the currently listed SGX pkg maintainers are either no longer working on SGX or unresponsive (no problem with either ofc). Wondering what we can do to get these PRs merged. Sadly full testing requires hardware or setting up a cloud VM, both of which are quite painful...

Any thoughts @sbellem?

File: pkgs/os-specific/linux/sgx/sdk/ipp-crypto.nix [hunk 1]
```diff
@@ -1,35 +1,29 @@
...
-      --replace 'ippcpo-config.cmake' 'ippcp-config.cmake'
-  '';
-
   cmakeFlags = [ "-DARCH=intel64" ] ++ extraCmakeFlags;
 
   nativeBuildInputs = [
     cmake
     nasm
+    # TODO(phlip9): The most recent `ipp-crypto` version (2021.9) now uses
+    # OpenSSL v3 so we can avoid this EOL version when sgx-sdk updates.
```
[the remark this answers is not in the document]
Comment (phlip9) [reply pkgs/os-specific/linux/sgx/sdk/ipp-crypto.nix]: Sorry for the delay! I went ahead and tested with `ipp-crypto` v2021.9.0 -- looks like everything is still fully functional. With this change we can finally remove the EOL OpenSSL version : )

**Comment by phlip9** [comment]
I went ahead and rebased the first 4 commits on a recent NixOS unstable -- no changes there. Following @arcz's suggestion I've also updated `ipp-crypto` from 2021.7 -> 2021.9.0, which allows us to remove the dependency on EOL OpenSSL v1.1!

After testing on my Azure gen2 NixOS VM ([DCsv3](https://learn.microsoft.com/en-us/azure/virtual-machines/dcv3-series)) I can confirm that everything appears to be in working order. This includes the NixOS `aesmd` service; I think it failed previously due to a minor misconfiguration issue on my part...

File: pkgs/os-specific/linux/sgx/ssl/default.nix [hunk 2]
```diff
@@ -1,44 +1,39 @@
...
 in
-stdenv.mkDerivation rec {
+stdenv.mkDerivation {
   pname = "sgx-ssl" + lib.optionalString debug "-debug";
   version = "${sgxVersion}_${opensslVersion}";
 
   src = fetchFromGitHub {
     owner = "intel";
     repo = "intel-sgx-ssl";
     rev = "lin_${sgxVersion}_${opensslVersion}";
```
Comment (blitz) [inline pkgs/os-specific/linux/sgx/ssl/default.nix:22]: They also have a [`3.0_Rev2`](https://github.com/intel/intel-sgx-ssl/tree/3.0_Rev2) version that builds against OpenSSL 3.0.

I'm not sure what the implications are for users of sgx-ssl, but given that OpenSSL 1.1 is EOL, it might be good to get rid of it here as well, if possible.

This can be another PR though.

File: pkgs/os-specific/linux/sgx/psw/default.nix [hunk 3]
```diff
@@ -131,6 +131,31 @@ stdenv.mkDerivation rec {
...
 
+  stripDebugList = [
+    # Only strip the host libraries. If we don't strip, we accidentally pull in
+    # `gcc`...
+    #
+    # We need to avoid stripping the enclave ".signed.so" files, since that
+    # would modify the enclave measurement (hash).
+    #
+    # TODO(phlip9): put the enclave libraries in a separate directory or tell
+    # strip to ignore all "*.signed.so".
```
Comment (blitz) [inline pkgs/os-specific/linux/sgx/psw/default.nix:142]: Is there a way to make this derivation fail to build when we fail to strip a binary here?

**Review by blitz — APPROVED** [review]
Thanks for updating this!

I've rebased this locally to master as of today and built the NixOS aesmd test. Looks good. Unfortunately, I don't have a real SGX system anymore.

**Review by RaitoBezarius — APPROVED** [review]
I don't have SGX hardware (well, except if I hack a bit my laptop who didn't upgrade their microcode to disable SGX) so I cannot test it but the diff looks good to me, thank you for taking care of that.

**Comment by blitz** [comment]
@sbellem @phlip9 If you are happy with this PR, I'd suggest merging it directly and skipping/closing #236815 for simplicity.

File: pkgs/os-specific/linux/sgx/psw/default.nix [hunk 4]
```diff
@@ -75,8 +76,7 @@ stdenv.mkDerivation rec {
 
   dontUseCmakeConfigure = true;
 
-  # Randomly fails if enabled
-  enableParallelBuilding = false;
+  enableParallelBuilding = true;
```
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/sgx/psw/default.nix:79]: That's set by default by cmake, so it is redundant to set it to true here. Please remove

File: pkgs/os-specific/linux/sgx/psw/default.nix (hunk 3 above)
Comment (SuperSandro2000) [reply pkgs/os-specific/linux/sgx/psw/default.nix:142]: Or alternatively: could we copy the files in after the stripping?

File: pkgs/os-specific/linux/sgx/sdk/ipp-crypto.nix [hunk 5]
```diff
@@ -1,36 +1,30 @@
...
-    hash = "sha256-QEJXvQ//zhQqibFxXwPMdS1MHewgyb24LRmkycVSGrM=";
+    hash = "sha256-+ITnxyrkDQp4xRa+PVzXdYsSkI5sMNwQGfGU+lFJ6co=";
   };
 
-  # Fix typo: https://github.com/intel/ipp-crypto/pull/33
-  postPatch = ''
-    substituteInPlace sources/cmake/ippcp-gen-config.cmake \
-      --replace 'ippcpo-config.cmake' 'ippcp-config.cmake'
-  '';
+  enableParallelBuilding = true;
```
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/sgx/sdk/ipp-crypto.nix:20]: Same here

File: pkgs/os-specific/linux/sgx/ssl/default.nix (hunk 2 above)
Comment (phlip9) [reply pkgs/os-specific/linux/sgx/ssl/default.nix:22]: Good catch! I'll take a look at this in a follow-up PR 👍

File: pkgs/os-specific/linux/sgx/psw/default.nix (hunk 3 above)
Comment (phlip9) [reply pkgs/os-specific/linux/sgx/psw/default.nix:142]: Went with the second approach here: move the enclaves to a $TMPDIR during fixup, then move them back afterwards. Should be significantly more robust.

**Comment by phlip9** [comment]
Ok everything should be good to go! I've (1) rebased on master, (2) removed the redundant `enableParallelBuilding = true;`, and (3) moved to a more robust approach for stripping `sgx-psw`, where we temporarily move the enclaves out before stripping, then move them back in `postFixup`.

**Comment by sbellem** [comment]
> @sbellem @phlip9 If you are happy with this PR, I'd suggest merging it directly and skipping/closing #236815 for simplicity.

All good! Closed the PR!
